### ISSUE TITLE
Inefficient DynamicGraph causes hang with only 100 tests.

### ISSUE THREAD

**Issue by steveprentice** [cbeust/testng#1710]
### TestNG Version
6.14.3, but behavior introduced with 6.12.

### Expected behavior

DynamicGraph doesn't spin endlessly in `setStatus()`

### Actual behavior

DynamicGraph has a few very hot spots where there are many levels of nested for-loops and where the algorithm complexity is O(n^4) or worse. This causes what feels like a hang, but is actually just `setStatus()` creating millions and millions of edges. This behavior was introduced in 015a9312ea8407902d5e1bcecf100d39db179751. Removing that code resolves this issue, but obviously causes a regression that that commit is meant to fix.

Below is the result of profiling this code.
![image](https://user-images.githubusercontent.com/81890/36867752-f483278c-1d4a-11e8-9cd6-b30c1ca972c1.png)

I've given quite a bit of thought to a potential fix, but struggling to balance DynamicGraphTest#test3 and DynamicGraphTest#edgeWeightTest1 without making fairly sizeable changes to DynamicGraph in general. The approach that I think would be best would be to make DynamicGraph only deal with true dependencies and then add sorting/chunking to `getFreeNodes()` or a layer above it. Doing it in this way makes it so that DynamicGraph doesn't need to deal with cycles / circular dependencies at all (other than preventing them). Right now, DynamicGraph is in an odd place where it prevents some, but explicitly allows and ignores others and then it later has to deal with odd corner cases such as the one introduced in the commit referenced above.

For example, if DynamicGraph only represented dependencies, for the graph for DynamicGraphTest#test3 would look like: a -> b, c -> b. (This is similar to how m_edges represents this graph for this test, but m_allEdges shows the other b -> a edge at a lower weight). `getFreeNodes()` would immediately be able to return that b is the only free node. Once that is done, b and its edges are removed from the graph and now a and c are free. The sorting / chunking layer would then chunk that into two chunks of work because a and c are different priorities. If we are preserving order between these nodes, we'd similarly chunk the work.

### Is the issue reproducible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

### Test case sample
The below test case reproduces the issue. It seems a bit contrived, but it's at least fairly concise. In our real-world example, we have about 200 tests and use priorities 1-4 on about 20 of those tests and the rest have the default priority. I couldn't reproduce with just using priority in a single test class however. My assumption is that other edges get added to the graph because in our real-world example, the tests are spread across files. In our real case, passing in `-parallel classes -threadcount 1` works around the issue – I assume because many `preserveOrder` edges are not added. In the case below, passing those options do not work around the issue, presumably because the edges are mostly based on priority and dependsOn which are not reduced during parallel execution.

```java
public class TestPriority {
    @Test(priority = 1, groups = { "group1" }) void test_1_1() { }
    @Test(priority = 2, groups = { "group1" }) void test_1_2() { }
    @Test(priority = 3, groups = { "group1" }) void test_1_3() { }
    @Test(priority = 1, groups = { "group1" }) void test_1_4() { }
    @Test(priority = 2, groups = { "group1" }) void test_1_5() { }
    @Test(priority = 6, groups = { "group1" }) void test_1_6() { }
    @Test(priority = 7, groups = { "group1" }) void test_1_7() { }
    @Test(priority = 8, groups = { "group1" }) void test_1_8() { }
    @Test(priority = 9, groups = { "group1" }) void test_1_9() { }
    @Test(priority = 10, groups = { "group1" }) void test_1_10() { }
    @Test(priority = 1, groups = { "group1" }) void test_2_1() { }
    @Test(priority = 2, groups = { "group1" }) void test_2_2() { }
    @Test(priority = 3, groups = { "group1" }) void test_2_3() { }
    @Test(priority = 1, groups = { "group1" }) void test_2_4() { }
    @Test(priority = 2, groups = { "group1" }) void test_2_5() { }
    @Test(priority = 6, groups = { "group1" }) void test_2_6() { }
    @Test(priority = 7, groups = { "group1" }) void test_2_7() { }
    @Test(priority = 8, groups = { "group1" }) void test_2_8() { }
    @Test(priority = 9, dependsOnGroups = { "group1" }) void test_3_9() { }
    @Test(priority = 10, dependsOnGroups = { "group1" }) void test_3_10() { }
    @Test(priority = 1, dependsOnGroups = { "group1" }) void test_4_1() { }
    @Test(priority = 2, dependsOnGroups = { "group1" }) void test_4_2() { }
    @Test(priority = 3, dependsOnGroups = { "group1" }) void test_4_3() { }
    @Test(priority = 1, dependsOnGroups = { "group1" }) void test_4_4() { }
    @Test(priority = 2, dependsOnGroups = { "group1" }) void test_4_5() { }
    @Test(priority = 6, dependsOnGroups = { "group1" }) void test_4_6() { }
    @Test(priority = 7, dependsOnGroups = { "group1" }) void test_4_7() { }
    @Test(priority = 8, dependsOnGroups = { "group1" }) void test_4_8() { }
    @Test(priority = 9, dependsOnGroups = { "group1" }) void test_4_9() { }
    @Test(priority = 10, dependsOnGroups = { "group1" }) void test_4_10() { }
    @Test(priority = 1, dependsOnGroups = { "group1" }) void test_5_1() { }
    @Test(priority = 2, dependsOnGroups = { "group1" }) void test_5_2() { }
    @Test(priority = 3, dependsOnGroups = { "group1" }) void test_5_3() { }
    @Test(priority = 1, dependsOnGroups = { "group1" }) void test_5_4() { }
    @Test(priority = 2, dependsOnGroups = { "group1" }) void test_5_5() { }
    @Test(priority = 6, dependsOnGroups = { "group1" }) void test_5_6() { }
    @Test(priority = 7, dependsOnGroups = { "group1" }) void test_5_7() { }
    @Test(priority = 8, dependsOnGroups = { "group1" }) void test_5_8() { }
    @Test(priority = 9, dependsOnGroups = { "group1" }) void test_5_9() { }
    @Test(priority = 10, dependsOnGroups = { "group1" }) void test_5_10() { }

    @Test(priority = 9) void test_2_9() { }
    @Test(priority = 10) void test_2_10() { }
    @Test(priority = 1) void test_3_1() { }
    @Test(priority = 2) void test_3_2() { }
    @Test(priority = 3) void test_3_3() { }
    @Test(priority = 1) void test_3_4() { }
    @Test(priority = 2) void test_3_5() { }
    @Test(priority = 6) void test_3_6() { }
    @Test(priority = 7) void test_3_7() { }
    @Test(priority = 8) void test_3_8() { }
    @Test(priority = 1) void test_6_1() { }
    @Test(priority = 2) void test_6_2() { }
    @Test(priority = 3) void test_6_3() { }
    @Test(priority = 1) void test_6_4() { }
    @Test(priority = 2) void test_6_5() { }
    @Test(priority = 6) void test_6_6() { }
    @Test(priority = 7) void test_6_7() { }
    @Test(priority = 8) void test_6_8() { }
    @Test(priority = 9) void test_6_9() { }
    @Test(priority = 10) void test_6_10() { }
    @Test(priority = 1) void test_7_1() { }
    @Test(priority = 2) void test_7_2() { }
    @Test(priority = 3) void test_7_3() { }
    @Test(priority = 1) void test_7_4() { }
    @Test(priority = 2) void test_7_5() { }
    @Test(priority = 6) void test_7_6() { }
    @Test(priority = 7) void test_7_7() { }
    @Test(priority = 8) void test_7_8() { }
    @Test(priority = 9) void test_7_9() { }
    @Test(priority = 10) void test_7_10() { }
    @Test(priority = 1) void test_8_1() { }
    @Test(priority = 2) void test_8_2() { }
    @Test(priority = 3) void test_8_3() { }
    @Test(priority = 1) void test_8_4() { }
    @Test(priority = 2) void test_8_5() { }
    @Test(priority = 6) void test_8_6() { }
    @Test(priority = 7) void test_8_7() { }
    @Test(priority = 8) void test_8_8() { }
    @Test(priority = 9) void test_8_9() { }
    @Test(priority = 10) void test_8_10() { }
    @Test(priority = 1) void test_9_1() { }
    @Test(priority = 2) void test_9_2() { }
    @Test(priority = 3) void test_9_3() { }
    @Test(priority = 1) void test_9_4() { }
    @Test(priority = 2) void test_9_5() { }
    @Test(priority = 6) void test_9_6() { }
    @Test(priority = 7) void test_9_7() { }
    @Test(priority = 8) void test_9_8() { }
    @Test(priority = 9) void test_9_9() { }
    @Test(priority = 10) void test_9_10() { }
    @Test(priority = 1) void test_10_1() { }
    @Test(priority = 2) void test_10_2() { }
    @Test(priority = 3) void test_10_3() { }
    @Test(priority = 1) void test_10_4() { }
    @Test(priority = 2) void test_10_5() { }
    @Test(priority = 6) void test_10_6() { }
    @Test(priority = 7) void test_10_7() { }
    @Test(priority = 8) void test_10_8() { }
    @Test(priority = 9) void test_10_9() { }
    @Test(priority = 10) void test_10_10() { }
}
```

**Comment by cbeust** [comment]
That's a fantastic analysis, Steve, thanks!

@juherr, @krmahadevan, any thoughts?

**Comment by krmahadevan** [comment]
@cbeust - I agree. But the `DynamicGraph` is one portion of TestNG that I haven't had too much experience with.  But yes in general I think we need to take a closer look at `DynamicGraph` to see areas of bottleneck in it (I remember seeing issues around it in the past as well).

@steveprentice - since you have spent so much time analyzing this, perhaps you could also propose a fix via a PR ?

**Comment by steveprentice** [comment]
@krmahadevan, I certainly don't mind opening a PR, but before heading down that route, I'd like to get some first impressions on my proposed approach to the fix. I spent a bunch of time trying to reduce the algorithmic complexity, from a big O perspective, of DynamicGraph.setStatus, but I think I've reached the conclusion that it can't be reduced much without a small redesign and I didn't want to embark on that without some discussion.

**Comment by krmahadevan** [comment]
@steveprentice - Sure that makes sense. But I was of the opinion that a PR would be better poised to bring out the gist of changes and facilitate an easier discussion.

**Comment by steveprentice** [comment]
It's a rather large design change and it'll take an investment on my part. I'm hoping to get some initial thoughts before heading down that path. You, @cbeust and @juherr are more familiar with the internals, pitfalls and corner cases than I by a large margin, so that initial assessment is important to me. The proposed approach to a fix is brief at the moment, but I think it accurately describes the direction. DynamicGraph would become a non-weighted graph of concrete dependencies. Priorities and preserveOrders would be handled at a different layer.

**Comment by steveprentice** [comment]
I realize I've both characterized the proposed changes as a small and a large redesign. I guess my perspective changes depending on who's doing the work! 🤔

**Comment by juherr** [comment]
I'm not feeling it good is we remove some dependencies like priorities or preserveOrders because it was already done before and we had a lot of issues about an unpredictable order.
I don't think we should simplify the graph logic because it modelizes well what TestNG is doing.

But at the same time, the graph logic is complicated and its implementation is clearly not the best :)
Maybe we can improve the algorithm but I'm not a graph expert.
Another potential option I've never tried is using a 3rd party graph library.

**Comment by krmahadevan** [comment]
@juherr 

> I'm not feeling it good is we remove some dependencies like priorities or preserveOrders because it was already done before and we had a lot of issues about an unpredictable order.

I guess @steveprentice was trying to say that we should perhaps separate out the logic that deals with ordering from the graph construction logic such that, the ordering logic gets played over the graph construction.

**Comment by juherr** [comment]
But the graph is used to manage the ordering logic :/

**Comment by cbeust** [comment]
I feel that we have good enough test coverage that we would catch any regressions, does everyone agree?

As long as the test don't break, I'm open to any rewrite that will improve things, even using a third party graph library (although it might be challenging to add the various behaviors on ordering that TestNG supports).

**Comment by juherr** [comment]
Sure! Any tentative is a good idea :)
